Fix Mega chunking error when using decoder-only model

tanaymeh opened the pull request:
# What does this PR do?

This PR aims to fix the error caused by `MegaModel` when the `is_decoder` setting is used in conjunction with `use_chunking` and `chunk_size` settings.

The error is described in detail [here](https://github.com/huggingface/transformers/issues/23331#issuecomment-1693729295).

Fixes #23331

## Who can review?
@ArthurZucker

tanaymeh committed:
add: potential fix to mega chunking in decoder only model bug

ArthurZucker commented:
Hey! Feel free to ping me when the PR is ready for review 😉

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25765). All of your documentation changes will be reflected on that endpoint.

tanaymeh commented:
@ArthurZucker You may review it now!

ArthurZucker reviewed (commented):
Nice! Can you add a test in the testing suite? This way we can make sure the chunking mode is properly tested!

tanaymeh commented:
Will do. Two questions regarding this:
1. For the test of chunking, I should only check the expected sizes of output or something else?
2. Conceptually, do you see any troubles with this solution in its current state? I am a little suspicious because of how simple the solutions was 😅

ArthurZucker commented:
1. Yes you could check expected sizes on a smaller model (using the small configs) 
2. I don't really see a problem since the attention mask created use the newly defined sequence length, and the code was already very clean so probably juste a typo!

tanaymeh committed:
add: decoder with chunking test

tanaymeh commented:
Thanks for confirming @ArthurZucker. I have added a test that checks [here](https://github.com/tanaymeh/transformers/blob/bb6c5eaff2421766daa79d5b489318a9a2c54726/tests/models/mega/test_modeling_mega.py#L316-L340) if the `attentions` returned in the `CausalLMOutputWithCrossAttentions` have their last dimension (`shape[-1]`), which is supposed to be `sequence_length` is equal to the `chunk_size` or not.

If the checks I added in `modeling_mega.py` are correct, it will use `chunk_size` instead of the actual `sequence_length`.

Is this correct, or shall I do any changes?

ArthurZucker commented on tests/models/mega/test_modeling_mega.py:
```diff
@@ -313,6 +313,32 @@ def create_and_check_decoder_model_past_large_inputs(
+
+        model = MegaForCausalLM(config).to(torch_device).eval()
+
+        result = model(
+            input_ids.repeat(1, 8),
+        )
```
```suggestion
        result = model(input_ids.repeat(1, 8))
```
we coul include attention masks as well to make sure the logic holds for attention masks that are given as an input.

tanaymeh replied:
In that case, will the passed-in attention mask have the size `[bs, seq_len]` or `[bs, chunk_size]`?

ArthurZucker reviewed (approved):
This is nice! Thanks for the long due fix 😉

tanaymeh committed:
add: input_mask passed with input_ids

tanaymeh commented:
Added your suggested changes @ArthurZucker! With `input_mask`, the tests pass too.

ArthurZucker commented:
Perfect! If you can just rebase on main to make sure the CIs are green?

tanaymeh commented:
> Perfect! If you can just rebase on main to make sure the CIs are green?

Done @ArthurZucker!

ArthurZucker commented:
Congrats on the PR 🚀 thanks for fixing!

tanaymeh commented:
Thanks a lot for helping @ArthurZucker!
